src/docs/widgets.md

```diff
@@ -0,0 +1,157 @@
+
+If you would like to have a look at the example code, please use the [Theia extension generator](https://github.com/eclipse-theia/generator-theia-extension). Install the generator, select the “Widget” example and enter “MyWidget” as a name for the extension.
+
+## Contributing a Widget (a View)
+
+Contributing a widget, in our example a view consists of three components:
+
+<ul>
+<li>The actual <b>widget</b>, responsible for:
+    <ul>
+    <li>Setting base parameters, such as an ID, a label and an icon</li>
+    <li>Creating the actual UI and its behavior</li>
+    <li>Reacting to lifecycle events such as `onUpdateRequest` or `onResize`</li>
+    </ul>
+</li>
+<li>A <b>widget factory</b>, responsible for instantiating the widget</li>
+<li>A <b>widget contribution</b>, responsible for wiring the view with the Theia workbench so that the widget can be opened from within the Theia workbench, e.g. via the views menus</li>
+</ul>
+
+### Implementing a Widget
+
+For the implementation of custom widgets, Theia provides several base classes to inherit from. This enables you to focus on creating your custom UI only, as the base classes already implement most required functions. Theia does not depend on a specific UI technology, like React, Vue.js or Angular. However, it provides convenience support by providing respective base classes, e.g. for React. If in doubt, using React is the most common choice for implementing custom widgets at the moment. Below you find an excerpt of the class hierarchy. In case you want to implement a widget using react, choose `ReactWidget` as a base class. If you want to implement a widget that mainly displays a tree, use `TreeWidget`. If you do not want to use React, use `BaseWidget`. Browse the type hierarchy of `BaseWidget` to see other available options.
+
+<ul>
+<li>BaseWidget
+    <ul>
+    <li><b>ReactWidget</b>
+        <ul>
+        <li>TreeWidget</li>
+        <li>…</li>
+        </ul>
+    </li>
+    </ul>
+</li>
+</ul>
+
+In the following code examples, we use `ReactWidget` as a base class. As shown below, we first initialize the widget with some base parameters:
+
+* `id`: To uniquely identify the widget, e.g. to open it via the WidgetManager.
+* `label`: Shown in the tab when the widget is open.
+* `caption`: Shown when hovering over the tab when the widget is open.
+* `closable`: Whether the user can close the widget (via the “x” in the tab or via right click menu).
+* `iconClass`: The icon shown in the tab when the widget is opened.
+
+**mywidget-widget.ts**
+```typescript
+@injectable()
+export class MyWidget extends ReactWidget {
+
+static readonly ID = 'my:widget';
+static readonly LABEL = 'My Widget';
+
+@postConstruct()
+protected async init(): Promise < void> {
+    this.id = MyWidget.ID;
+    this.title.label = MyWidget.LABEL;
+    this.title.caption = MyWidget.LABEL;
+    this.title.closable = true;
+    this.title.iconClass = 'fa fa-window-maximize'; // example widget icon.
+    this.update();
```

Comment:
In most cases there is no difference between placing statements in the constructor or placing them in a "@postConstruct", as postconstuct is executed directly after the constructor. Therefore you can find both in the Theia codebase. The only exception is if the initialization needs to access a field injection, which would ot work in the constructor.

If you would like to use async/await it's preferable to use "@postConstruct" as the annotated method can be async. When a method is async and doesn't return anything then it's technically a Promise<void>. In this specific case however the "async" doesn't do anything, so we could remove the async and change it to a normal void return type. However the async also doesn't hurt here.

Generally speaking "@postConstruct" is probably preferable over constructors as this offers more flexibility in using the class. Especially it eases testing. If the logic is part of the constructor it can't be avoided or be executed in a controllable manner.

This is what you find often in the Theia codebase in widget classes: A constructor which only contains the configuration options needed for the Widget base class and a "@postConstruct"  containing the whole logic. I guess this is also the reason why it's done this way in the template.

So this does not really refer to the widgets lifecycle, it is more a coding style question. If you think we should generally discuss this, I would like to do it in a separate ticket, please. If we want to get rid of the async, we should also adapt the template.